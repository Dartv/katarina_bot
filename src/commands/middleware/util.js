import R from 'ramda';

import { lenses } from '../../util';
import { findByRef } from '../../util/helpers';
import { ErrorResponse } from '../responses';

export const isImageOwner = ({ user, image }) => R.eqProps('id', user, image.user);

export const isAdmin = ({ message: { member } }) => member.hasPermission('ADMINISTRATOR');

export const findByArgsRef = lookup => R.converge(findByRef, [
  R.view(lenses.args.ref),
  lookup,
]);
export const findGuildImageByArgsRef = findByArgsRef(R.view(lenses.guild.images));
export const findUserImageByArgsRef = findByArgsRef(R.view(lenses.user.images));

export const assignImage = setter => R.converge(R.set(lenses.image), [
  setter,
  R.identity,
]);
export const assignImageByGuildImage = assignImage(findGuildImageByArgsRef);
export const assignImageByUserImage = assignImage(findUserImageByArgsRef);

export const createErrorResponse = R.curry((lookup, messageCreator) => R.converge(ErrorResponse, [
  R.compose(messageCreator, lookup),
  R.identity,
]));
export const createErrorResponseFromArgsRef = createErrorResponse(R.view(lenses.args.ref));
