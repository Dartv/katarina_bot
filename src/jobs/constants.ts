export const DAILY_RESET_CRON = '0 6 * * *';
