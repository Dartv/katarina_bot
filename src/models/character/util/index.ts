export { default as createCharacterEmbed } from './createCharacterEmbed';
export { default as getCharacterStarRating } from './getCharacterStarRating';
