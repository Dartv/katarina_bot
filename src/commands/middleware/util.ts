import R from 'ramda';

import { lenses } from '../../util';
import { findByRef } from '../../util/helpers';
import { ErrorResponse } from '../responses';

export const isImageOwner = ({ user, image }) => R.eqProps('id', user, image.user);

export const isAdmin = ({ message: { member } }) => member.hasPermission('ADMINISTRATOR');

export const findByArgsRef = lookup => R.converge(findByRef, [
  R.view(lenses.args.ref as any),
  lookup,
]);
export const findGuildImageByArgsRef = findByArgsRef(R.view(lenses.guild.images as any));
export const findUserImageByArgsRef = findByArgsRef(R.view(lenses.user.images as any));

export const assignImage = setter => R.converge(R.set(lenses.image), [
  setter,
  R.identity,
]);
export const assignImageByGuildImage = assignImage(findGuildImageByArgsRef);
export const assignImageByUserImage = assignImage(findUserImageByArgsRef);

export const createErrorResponse = R.curry((lookup, messageCreator) => R.converge(ErrorResponse, [
  R.compose(messageCreator, lookup),
  R.identity,
]));
export const createErrorResponseFromArgsRef = createErrorResponse(R.view(lenses.args.ref as any));

export const composeMiddleware = (...functions) => {
  if (functions.length === 1) {
    return functions[0];
  }

  return functions.reduceRight((f, next) => (...args) => next(f, ...args));
};

export function applyMiddleware(...middleware) {
  middleware.forEach((layer) => {
    if (typeof layer !== 'function') {
      throw new TypeError('Expected all provided middleware to be functions.');
    }
  });

  return (handler) => {
    if (typeof handler !== 'function') {
      // eslint-disable-next-line max-len
      throw new TypeError('Expected handler to be a function. Middleware can only be applied to functions.');
    }

    return composeMiddleware(...middleware, handler);
  };
}

export const combineMiddleware = <T extends (...args: any[]) => any>(
  ...middleware: T[]
) => (next: T, ...args: any[]): T => applyMiddleware(...middleware)(next)(...args);
