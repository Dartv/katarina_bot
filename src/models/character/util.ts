import { CharacterStar } from '../../util';

export function getCharacterStarRating(popularity: number): CharacterStar {
  if (popularity <= 500) return CharacterStar.FIVE_STAR;
  if (popularity <= 3000) return CharacterStar.FOUR_STAR;
  if (popularity <= 10000) return CharacterStar.THREE_STAR;
  return CharacterStar.TWO_STAR;
}
