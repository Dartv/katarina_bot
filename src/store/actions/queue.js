import R from 'ramda';

import { lenses } from '../../util';

const initialState = { entities: {}, allIds: [] };
const entity = id => R.compose(
  lenses.entities,
  R.lensProp(id),
);

const viewOr = R.curryN(3,
  (defaultValue, lens, data) => R.defaultTo(defaultValue, R.view(lens, data))
);

const reduceQueue = R.curry((reducer, id, state) => ({
  ...state,
  entities: {
    ...state.entities,
    [id]: reducer(state.entities[id] || initialState),
  },
}));

export const enqueue = (item, id) => reduceQueue(R.compose(
  R.over(lenses.allIds, R.append(item.id)),
  R.set(R.compose(lenses.entities, R.lensProp(item.id)), item),
), id);


export const dequeue = reduceQueue((state) => {
  const id = R.view(lenses.allIds.head);
  return R.when(
    id,
    R.compose(
      R.over(lenses.allIds, R.drop(1)),
      R.over(lenses.entities, R.dissoc(id(state))),
    ),
    state,
  );
});

export const clear = reduceQueue(R.compose(
  R.set(lenses.entities, {}),
  R.set(lenses.allIds, []),
));

export const getQueue = context => state => () =>
  viewOr(initialState, entity(context.message.guild.id), state);

export const peek = context => () => dispatch => R.compose(
  R.converge(R.view, [
    R.compose(entity, R.view(lenses.allIds.head)),
    R.identity,
  ]),
  R.compose(dispatch, getQueue),
)(context);

export const getQueueItem = (id, context) => () => dispatch => R.compose(
  R.view(entity(id)),
  R.compose(dispatch, getQueue),
)(context);

export const getQueueSize = context => () => dispatch => R.compose(
  R.length,
  R.view(lenses.allIds),
  R.compose(dispatch, getQueue),
)(context);
