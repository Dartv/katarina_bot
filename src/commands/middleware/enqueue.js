import { ErrorResponse, SuccessResponse } from '../responses';
import { getQueueItem, enqueue } from '../../store/actions/queue';
import { ERRORS } from '../../util/constants';


export default () => async (next, context) => {
  const { services, message: { guild }, video, dispatch, formatter: { bold } } = context;
  const store = services.get('music.store');

  if (!guild.voiceConnection) return new ErrorResponse(ERRORS.VC_NOT_FOUND, context);

  const queueItem = store.dispatch(getQueueItem(video.id, context));

  if (queueItem) return new ErrorResponse(ERRORS.VC_ALREADY_QUEUED, context);

  store.dispatch(enqueue(video, guild.id));

  await dispatch(new SuccessResponse(`Successfully queued ${bold(video.title)}`, '', context));

  return next(context);
};
