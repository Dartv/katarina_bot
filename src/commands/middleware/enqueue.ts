import { ErrorResponse, SuccessResponse } from '../responses';
import { getQueueItem, enqueue } from '../../store/actions/queue';
import { ERRORS } from '../../util/constants';


export default () => async (next, context) => {
  const { message: { guild } } = context;

  const {
    services, video, dispatch, formatter: { bold },
  } = context;
  const store = services.get('music.store');
  const queueItem = store.dispatch(getQueueItem(video.id, context));

  if (queueItem) return ErrorResponse(ERRORS.YT_ALREADY_QUEUED, context);

  store.dispatch(enqueue(video, guild.id));

  await dispatch(SuccessResponse(`Successfully queued ${bold(video.title)}`, '', context));

  return next(context);
};
