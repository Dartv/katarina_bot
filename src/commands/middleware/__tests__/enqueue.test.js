import R from 'ramda';
import Store, { thunk } from 'repatch';

import { createContext } from '../../../util/tests';
import { ERRORS } from '../../../util/constants';
import enqueue from '../enqueue';
import { getQueueItem } from '../../../store/actions/queue';

jest.mock('../../responses/ErrorResponse');

describe('enqueue', () => {
  let store;
  const next = R.T;
  const getContext = state => createContext({
    message: {
      guild: {
        id: 1,
        voiceConnection: {},
      },
    },
    services: {
      get: () => {
        store = new Store(state).addMiddleware(thunk);
        return store;
      },
    },
  });
  const state = {
    entities: {
      1: {
        entities: {
          2: {
            id: 2,
          },
        },
        allIds: [2],
      },
    },
    allIds: [1],
  };

  beforeEach(() => {
    store = null;
  });

  it('should match snapshots', () => {
    expect(state).toMatchSnapshot();
    expect(getContext(state)).toMatchSnapshot();
  });

  it('should dispatch an error if no voice channel found', async () => {
    const context = {
      message: {
        guild: {
          voiceConnection: null,
        },
      },
    };
    const { executor } = await enqueue()(next, context);
    const response = await executor(context);

    expect(response).toBe(ERRORS.VC_NOT_FOUND);
  });

  it('should dispatch an error if the item is already queued', async () => {
    const context = {
      ...getContext(state),
      video: {
        id: 2,
      },
    };
    const { executor } = await enqueue()(next, context);
    const response = await executor(context);

    expect(response).toBe(ERRORS.YT_ALREADY_QUEUED);
  });

  it('should pass if successfully queued', async () => {
    const video = { id: 3 };
    const context = {
      ...getContext(state),
      video,
    };
    const response = await enqueue()(next, context);
    const item = store.dispatch(getQueueItem(video.id, context));

    expect(response).toBe(true);
    expect(item).toEqual(video);
  });
});
