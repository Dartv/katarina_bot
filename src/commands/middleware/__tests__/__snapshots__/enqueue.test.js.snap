// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enqueue should match snapshots 1`] = `
Object {
  "allIds": Array [
    1,
  ],
  "entities": Object {
    "1": Object {
      "allIds": Array [
        2,
      ],
      "entities": Object {
        "2": Object {
          "id": 2,
        },
      },
    },
  },
}
`;

exports[`enqueue should match snapshots 2`] = `
Object {
  "args": Object {},
  "dispatch": [Function],
  "formatter": [Function],
  "message": Object {
    "author": Object {
      "avatarURL": "http://example.com",
      "discriminator": "1234",
      "username": "John",
    },
    "channel": Object {
      "send": [Function],
    },
    "guild": Object {
      "id": 1,
      "voiceConnection": Object {
        "channel": Object {},
      },
    },
    "member": Object {
      "voiceChannel": Object {},
      "voiceChannelID": null,
    },
  },
  "services": Object {
    "get": [Function],
  },
}
`;
